lib/Conversion/TorchToTosa/TorchToTosa.cpp

```diff
@@ -3028,6 +3028,50 @@ LogicalResult ConvertAtenOp<AtenWhereSelfOp>::matchAndRewrite(
 }
 
 
+template <>
+LogicalResult ConvertAtenOp<AtenSoftmaxIntOp>::matchAndRewrite(
```

Comment:
I think there is still more work to be done before I would be comfortable merging this PR. The first reason is that this creates a precedent of allowing per-op custom patterns for converting to `tosa::CustomOp`. Other contributors of torch-mlir who might need a different op lowered to `tosa::CustomOp` might follow this approach, and we definitely don't want future PRs adding more per-op custom patterns to `tosa::CustomOp`.

The second reason is that this is not yet a solution that can be generalized nicely to other ops. Two main reasons come to mind at the moment:

1.  How would this approach work if there was also a conversion pattern for `AtenSoftmaxIntOp` that converts to standard tosa ops? Which pattern would get applied when both patterns are active? This is where having a separate pass for custom ops simplifies things.

2. This does not address one of the issues raised in the RFC, which is the handling of types such as `!torch.none`. This PR simply ignores the third argument of `AtenSoftmaxIntOp`. If users start depending on this convention for this op, it will make it harder to change things in the future. We need to first come up with a good design for the rules of conversion for all types that users can then depend on. Essentially, we need to clearly define a `tosa::CustomOp` contract. For example, are we always going to convert constant `int`s to rank-1 tensors as is done in this PR, or should we use rank-0 tensors or encode the int as an attribute? This is crucial to get right in the beginning so that users can have something stable to depend on.

I think we should continue iterating on this.

Reply:
Can we split it into small pieces with several PR? We are not trying to implement everything of https://github.com/llvm/torch-mlir/issues/1519 in one PR.

Reply:
> I think there is still more work to be done before I would be comfortable merging this PR. The first reason is that this creates a precedent of allowing per-op custom patterns for converting to `tosa::CustomOp`. Other contributors of torch-mlir who might need a different op lowered to `tosa::CustomOp` might follow this approach, and we definitely don't want future PRs adding more per-op custom patterns to `tosa::CustomOp`.
> 
> The second reason is that this is not yet a solution that can be generalized nicely to other ops. Two main reasons come to mind at the moment:
> 
> 1. How would this approach work if there was also a conversion pattern for `AtenSoftmaxIntOp` that converts to standard tosa ops? Which pattern would get applied when both patterns are active? This is where having a separate pass for custom ops simplifies things.
> 2. This does not address one of the issues raised in the RFC, which is the handling of types such as `!torch.none`. This PR simply ignores the third argument of `AtenSoftmaxIntOp`. If users start depending on this convention for this op, it will make it harder to change things in the future. We need to first come up with a good design for the rules of conversion for all types that users can then depend on. Essentially, we need to clearly define a `tosa::CustomOp` contract. For example, are we always going to convert constant `int`s to rank-1 tensors as is done in this PR, or should we use rank-0 tensors or encode the int as an attribute? This is crucial to get right in the beginning so that users can have something stable to depend on.
> 
> I think we should continue iterating on this.

Can we split it into small pieces with several PR? We are not trying to implement everything of https://github.com/llvm/torch-mlir/issues/1519 in one PR.

Reply:
Can we split it into small pieces with several PR? We are not trying to implement everything of https://github.com/llvm/torch-mlir/issues/1519 in one PR.

Reply:
Sure! It's fine to implement in chunks with unimplemented paths resulting in an error with a good error message. Given that the first PR will lay down the structure that future PRs will build on, for this RFC the first PR should include a document describing the `tosa::CustomOp` contract as well as create the `TorchToTosaCustom` pass, and have a simple working example of an op being systematically converted to `tosa::CustomOp`.

Reply:
Cool, I just create a new ConvertTorchToTosaCustom pass and add the document in the description. The example is also there.